Villager Registry Changes

Thutmose opened the pull request:
If a profession is removed, any villager with that profession has levels
and trades reset, and is reset to farmer.
If a profession's id is changed, it automatically re-assigns it
accordingly, so professions can be added or removed as needed, with
trades only changing if the villager's profession is removed.

Setting a villager's profession will reset the career (i.e. randomize it) as well as resetting the trade list.  This does not affect reading from NBT, as profession is set before the other values are read.

Vanilla spawning works as it used to, with any building/spawnegg that
used to make random villagers being able to have the new ones as
well, but any building with specific villagers, still only makes that
specific villager, this is ensured so long as the new villagers are
registered after the vanilla ones.

The test mod adds three new professions, and it was tested by randomly
commenting some of them out, then running minecraft, then commenting
some of them back in.  The villagers which were not removed properly
kept their professions, and the ones that were removed were set to
farmers.  with the test mod, change the int toSet in the interact event handler to whichever villager profession you want it to set, then shift right click the villager with a stick, it will change the profession.

cpw commented:
How does it handle mismatches between client and server?

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -133,8 +136,7 @@ public void registerVillagerId(int id)
-    @Deprecated // Doesn't work at all.
-    public void registerVillagerSkin(int villagerId, ResourceLocation villagerSkin)
+    private void registerVillagerSkin(int villagerId, ResourceLocation villagerSkin)
     {
```
This is deprecated to be removed in 1.8.8 but still exists because we dont want to fire MissingMethodExceptions.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -17,23 +17,189 @@
  
-+            //TODO: Hook into VillagerRegistry to get name
+-            String s1 = null;
+-
```
This patch segment needs to be a lot smaller, no need to remove the switch if you just override it down below.

Thutmose replied:
ok, will not remove that block

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -17,23 +17,189 @@
+         {
+-            this.field_175563_bv = this.field_70146_Z.nextInt(aitradelist.length) + 1;
++            int careerCount = net.minecraftforge.fml.common.registry.VillagerRegistry.getNumberOfCareers(this.func_70946_n());
++            this.field_175563_bv = this.field_70146_Z.nextInt(careerCount) + 1;
```
Any resons why you go the route of using intermediate variables here? makes the patch file harder to read in my opinion.

Thutmose replied:
that particular case the intermediate variable was to avoid having an extremely long line, I will avoid doing that in the future.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -17,23 +17,189 @@
+     }
+ 
++    public void setCareer(int careerId)
++    {
```
New functions at end of file! Also, if this is JUST a field change then use a access transformer. But you should probably validate the career level and trades if you call this function.

Thutmose replied:
I will move it to end of file.  for validating, it would require essentially a modification of the populateBuyLists method inside the file (as that only checks if both career and level are zero), I can add that though.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -17,23 +17,189 @@
+     {
+-        return Math.max(this.field_70180_af.func_75679_c(16) % 5, 0);
++        return Math.max(this.field_70180_af.func_75679_c(16) % net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession.nextId, 0);
+     }
```
This isnt this simple as villager ids are NOT guaranteed to be densely packed in the registry, so it may be {farmer, blacksmith, null, beekeeper}

Thutmose replied:
the way I had changed it with the VillagerProfession incrementing nextId every time a profession is added, it is guarenteed to be densely packed.  I will need to see what needs to be changed when I move it from using hashmaps to the namespacedregistrty.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -17,23 +17,189 @@
++            this.field_175563_bv = 0;
++            this.field_175562_bw = 0;
++            this.field_70963_i = null;
++            this.func_70938_b(func_70946_n());
```
You duplicate code here, you clear things in setProfession but also clear them here...

Thutmose replied:
I will remove the extra clearings from there.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -17,23 +17,189 @@
+-        if (p_70037_1_.func_150297_b("Offers", 10))
++        
++        String newProfessionName = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForId(func_70946_n()).getName();
++        if(professionId != func_70946_n()||!(professionName.equals(newProfessionName)||professionName.isEmpty()))
```
check your spacing.

Thutmose replied:
I do not see where the potential NPE is in that area, the getProfessionForName method will never return null, if it was going to, it instead returns the value for farmer.  I will see about fixing the spacing.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -17,23 +17,189 @@
++        net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession profession = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForName(professionName);
++        String oldProfession = profession.getName();
++        if(oldProfession.equals(professionName) && professionId != profession.getId())
++        {
```
Spacing again, follow the format in the rest of the code.

Thutmose replied:
ok, for patches I will omit {} for one line if statements.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -17,23 +17,189 @@
+         super.func_70037_a(p_70037_1_);
+-        this.func_70938_b(p_70037_1_.func_74762_e("Profession"));
++        int professionId = p_70037_1_.func_74762_e("Profession");
++        String professionName = p_70037_1_.func_74779_i("professionName");
```
A hasKey check here is important else NPEs are probably gunna happen.

Thutmose replied:
Vanilla returns and empty string if you request a string that is not in the tag, I could add a hasKey check incase that changes in the future.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -17,23 +17,189 @@
+         super.func_70014_b(p_70014_1_);
+         p_70014_1_.func_74768_a("Profession", this.func_70946_n());
++        String professionName = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForId(this.func_70946_n()).getName();
++        p_70014_1_.func_74778_a("professionName", professionName);
```
Capitol P to follow the standard of the other names.

Thutmose replied:
will change that

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -160,7 +162,6 @@ public void registerVillageCreationHandler(IVillageCreationHandler handler)
      * @param defaultSkin
      */
     @SideOnly(Side.CLIENT)
-    @Deprecated // Doesn't work at all.
```
No keep it deprecated we do NOT want to encourage the usage of int based ids. Better to remove it in 1.8.8

Thutmose replied:
should I add a replacement method which takes the VillagerProfession and the texture instead?

LexManos replied:
The profession should have a texture location in it already so this function shouldn't be necessary at all.

Thutmose replied:
it is currently used by the renderer to get the texture, should I also make a patch for that to use the one from the profession?

Thutmose replied:
now just the one method that takes an int and a default texture is left, that one is currently used on line 51 of the RenderVillager class, but shouldn't be needed by any mods.

Thutmose replied:
It is still marked as deprecated, I just removed the comment about it doing nothing, as it is actually used by the RenderVillager

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -203,12 +204,23 @@ public void register(VillagerProfession prof)
-
+//    private FMLControlledNamespacedRegistry<VillagerProfession> professions = GameData.createRegistry("villagerprofessions", VillagerProfession.class, 0, 1024);
+    private HashMap<Integer, VillagerProfession> professionsById = new HashMap<Integer, VillagerProfession>();
+    private HashMap<String, VillagerProfession> professionsByName = new HashMap<String, VillagerProfession>();
```
No, just no, FMLControlledNamespacedRegistry is exactly this. A {Object, String, Int} -> Object

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -269,6 +281,31 @@ private void register(VillagerCareer career)
+
+        private void setId(int id)
+        {
+            if(id < nextId)
```
Should never need to know its int id.

Thutmose replied:
I was using this to keep track of the number added, as well as to keep the vanilla compatibility, since vanilla was using int ids for everything , I guess I could instead check the number of professions registered, then just use registry lookups instead.

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -250,16 +262,16 @@ private void init()
+        private int id;
         private List<VillagerCareer> careers = Lists.newArrayList();
-        private RegistryDelegate<VillagerProfession> delegate = GameData.getRegistry("villagerprofessions", VillagerProfession.class).getDelegate(this, VillagerProfession.class);
 
```
These are hever for important reasons stick with RamespacedRegistry

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -285,9 +324,46 @@ public VillagerCareer(VillagerProfession parent, String name)
+            {
+                ArrayList<ITradeList> tradesForLevel = new ArrayList<ITradeList>();
+                for(ITradeList trade: traids[i])
+                {
```
Protip, arraylist has a constructor that takes and array, also all lists have a addAll function.

Thutmose replied:
will swap over to using the array taking constructor

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -306,10 +392,54 @@ public boolean equals(Object o)
-        //TODO: Grab id range from internal registry
-        entity.setProfession(rand.nextInt(5));
+        VillagerProfession profession = INSTANCE.professionsByName.get(entries.get(prof));
+        entity.setProfession(profession.id);
```
Problem with this is that HashMaps are not ordered and since this is used in world gen things will most likely derp and you dont keep vanilla world gen compatibility, you need a ordered list of professions by ID.

Thutmose replied:
vanilla manually sets the profession IDs for any villagers which should be specific, then randomly picked any others.  this should just affect that set which is randomly picked, which is why I didn't think it will change anything, besides getting modded villager professions in the random villagers, should that not occur? I think preventing that from occuring will prevent the ability to breed for specific villagers as well.

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -306,10 +392,54 @@ public boolean equals(Object o)
+    {
+        VillagerProfession profession = INSTANCE.professionsById.get(professionID);
+        //The villagers seem to have career = 0 meaning no career.
+        if(careerId > profession.careers.size())
```
Your comment and if statement don't match, you're missing the <=0 test.

Thutmose replied:
that method is always called after the villager increments their career to 1 (or higher).  the way I wrote it, if the profession.careers is empty, then it will return with no trades added, as the value for careerId passed in is always > 0.

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -306,10 +392,54 @@ public boolean equals(Object o)
+    }
+    
+    public static VillagerProfession getProfessionForId(int professionID)
+    {
```
Should probably be marked deprecated/interal, we don't want modders accessing ANYTHING by int id.

Thutmose replied:
those methods were mainly intended for use inside the EntityVillager, to keep the vanilla compatibility, should I instead make them take the EntityVillager itself?

Thutmose commented:
I have made the changes requested by LexManos, I changed over to using the NamespacedRegistry, and made a getProfession method which takes the villager, and swapped over to mainly using it (I still needed the int one for the reading from nbt stuff, but added a comment that it should only be used internally).  I added a method for getting texture that takes the villager entity instead, so no IDs needed there.

as for how it handles server/client conflicts, here is what it looks like when one client is missing two of the added professions.  This is what they look like.

![image](https://cloud.githubusercontent.com/assets/2363472/11195055/1ebc53d2-8c86-11e5-87cb-b5124ad8e7ed.png)

but this is the gui given by the one that "looks" like a librarian:
![image](https://cloud.githubusercontent.com/assets/2363472/11195082/4f16634c-8c86-11e5-96dc-6b6aa400aa9a.png)

same gui as seen by the player not missing that profession:
![image](https://cloud.githubusercontent.com/assets/2363472/11195097/67f9c4da-8c86-11e5-8b29-38bb233fe824.png)
